Add docs for field splits, libmesh partitioners and for most testers

GiudGiud opened the pull request:
This PR shows a need for automated doc generation from Python classes. Anyone knows what we have for this?

GiudGiud committed:
Add parameter checks on setting off-diagonal couplings

GiudGiud committed:
Add docs for field split preconditioner, refs #18333 #17566 #16248 #15968

GiudGiud committed:
Move full and off-diagonal coupling preconditioner parameters to the base class since all current PCs implement them

Clang format

GiudGiud committed:
Add documentation for Split syntax for FSPs, refs #18333 #17566 #16248 #15968s

GiudGiud committed:
Add documentation to libmeshpartitioner, fix some typos, refs #18333 #17566 #16248 #15968

Describe partitioners in libmesh succintly

GiudGiud committed:
Add basic docs for all available testers & update their list, refs #19983

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/927439/) on accca33 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19985/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19985/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a6076d579c4ea10b0c3bdb7d7561334faf24c241`

GiudGiud committed:
Address test failures:
- flip default back to not full for preconditioners
- fix doc link
- fix up phase field inputs which want full and not full
Refs #15968

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/943047/) on 8cc3695 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19985/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19985/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2dab3bb92b96a65b8e31b2add8e8fc7891c55d6a`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/943394/) on 1c5c12e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19985/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/943627/) on 94963c5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2dab3b</th>
      <th colspan=4>#19985 94963c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.34%</td>
      <td>82.33%</td>
      <td>-0.01%</td>
      <td>74.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73444</td>
      <td>73447</td>
      <td>+3</td>
      <td>29</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15753</td>
      <td>15760</td>
      <td>+7</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19985/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19985/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19985/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 74.36% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented on framework/doc/content/source/preconditioners/FieldSplitPreconditioner.md:
```diff
@@ -1,33 +1,33 @@
 
-!! Replace these lines with information regarding the FSP object.
+The `FieldSplitPreconditioner` allows for custom preconditioning for each nonlinear variable in the numerical
+system. One or more variables may be targetted in a subsolve that will only consider part of the numerical system.
```
```suggestion
system. One or more variables may be targeted in a subsolve that will only consider part of the numerical system.
```

dschwen commented on framework/doc/content/source/splits/Split.md:
```diff
@@ -1,33 +1,22 @@
-
-## Example Input File Syntax
+This is an alternative syntax to the [FieldSplitPreconditioner.md].
+The field split based preconditioner is entirely set up by defining the relevant Petsc options.
```
```suggestion
The field split based preconditioner is entirely set up by defining the relevant PETSc options.
```

dschwen commented on framework/doc/content/source/splits/Split.md:
```diff
@@ -1,33 +1,22 @@
 
-!! Describe and include an example of how to use the Split object.
+More information about field split preconditioning may be found in the
+[Petsc manual]([here](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/).
```
```suggestion
[PETSc manual]([here](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/).
```

dschwen commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -30,22 +30,15 @@ FieldSplitPreconditioner::validParams()
                         true,
-                        "Set to true if you want the full set of couplings.  Simply "
-                        "for convenience so you don't have to set every off_diag_row "
+                        "Set to true if you want the full set of couplings between variables. "
+                        "Simply for convenience so you don't have to set every off_diag_row "
```
```suggestion
                        "Set to true if you want the full set of couplings between variables "
                        "simply for convenience so you don't have to set every off_diag_row "
```
this was not a sentence

dschwen commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -60,22 +53,24 @@ FieldSplitPreconditioner::FieldSplitPreconditioner(const InputParameters & param
     // off-diagonal entries
     std::vector<std::vector<unsigned int>> off_diag(n_vars);
-    for (unsigned int i = 0; i < getParam<std::vector<std::string>>("off_diag_row").size(); i++)
+    for (unsigned int i = 0; i < off_diag_rows.size(); i++)
```
might as well update this to
```suggestion
    for (const auto i: index_range(off_diag_rows))
```

dschwen commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -60,22 +53,24 @@ FieldSplitPreconditioner::FieldSplitPreconditioner(const InputParameters & param
   std::unique_ptr<CouplingMatrix> cm = std::make_unique<CouplingMatrix>(n_vars);
   if (!full)
   {
+    const auto off_diag_rows = getParam<std::vector<NonlinearVariableName>>("off_diag_row");
+    const auto off_diag_columns = getParam<std::vector<NonlinearVariableName>>("off_diag_column");
```
Is this enforcing both parameters to be of equal length? (probably in the hidden code). If so then a better way would be 

```
const auto off_diag = getParam<NonlinearVariableName, NonlinearVariableName>("off_diag_row", "off_diag_column");
```

to get a vector of pairs (with built in length error checking. Then you can simplify teh code below to 

```
for (const auto & item: off_diag)
{
  auto row = _nl.getVariable(0, item.first).number();
  auto column = _nl.getVariable(0, item.second).number();
  (*cm)(row, column) = 1;
}
```

dschwen commented on framework/src/preconditioners/FiniteDifferencePreconditioner.C:
```diff
@@ -83,12 +70,16 @@ FiniteDifferencePreconditioner::FiniteDifferencePreconditioner(const InputParame
     // off-diagonal entries
     std::vector<std::vector<unsigned int>> off_diag(n_vars);
-    for (unsigned int i = 0; i < getParam<std::vector<std::string>>("off_diag_row").size(); i++)
+    for (unsigned int i = 0;
+         i < getParam<std::vector<NonlinearVariableName>>("off_diag_row").size();
```
```suggestion
    for (const auto i: index_range(getParam<std::vector<NonlinearVariableName>>("off_diag_row")))
```
would be better here as it guarantees single evaluation of the loop bounds... and also looks more readable

dschwen commented on framework/src/preconditioners/FiniteDifferencePreconditioner.C:
```diff
@@ -83,12 +70,16 @@ FiniteDifferencePreconditioner::FiniteDifferencePreconditioner(const InputParame
     // off-diagonal entries
     std::vector<std::vector<unsigned int>> off_diag(n_vars);
-    for (unsigned int i = 0; i < getParam<std::vector<std::string>>("off_diag_row").size(); i++)
+    for (unsigned int i = 0;
+         i < getParam<std::vector<NonlinearVariableName>>("off_diag_row").size();
+         i++)
```
```suggestion
    for (const auto i: index_range(getParam<std::vector<NonlinearVariableName>>("off_diag_row")))
```
would be better here as it guarantees single evaluation of the loop bounds... and also looks more readable

dschwen commented on framework/src/preconditioners/MoosePreconditioner.C:
```diff
@@ -30,6 +30,20 @@ MoosePreconditioner::validParams()
 
+  params.addParam<std::vector<NonlinearVariableName>>(
+      "off_diag_row",
+      "The variable names for the off-diagonal rows you want to add into the matrix, they will be "
```
```suggestion
      "The variable names for the off-diagonal rows you want to add into the matrix; they will be "
```

dschwen commented on framework/src/preconditioners/MoosePreconditioner.C:
```diff
@@ -30,6 +30,20 @@ MoosePreconditioner::validParams()
+      "associated with an off-diagonal column from the same position in off_diag_column.");
+  params.addParam<std::vector<NonlinearVariableName>>(
+      "off_diag_column",
+      "The variable names for the off-diagonal columns you want to add into the matrix, they will "
```
```suggestion
      "The variable names for the off-diagonal columns you want to add into the matrix; they will "
```

dschwen commented on framework/src/preconditioners/MoosePreconditioner.C:
```diff
@@ -30,6 +30,20 @@ MoosePreconditioner::validParams()
+      "be associated with an off-diagonal row from the same position in off_diag_row.");
+  params.addParam<bool>("full",
+                        false,
+                        "Set to true if you want the full set of couplings between variables. "
+                        "Simply for convenience so you don't have to set every off_diag_row "
```
```suggestion
                        "Set to true if you want the full set of couplings between variables "
                        "simply for convenience so you don't have to set every off_diag_row "
```

dschwen commented on modules/doc/content/python/testers/Exodiff.md:
```diff
@@ -0,0 +1,41 @@
+
+- `map`: Use geometrical mapping to match up elements. This is usually a good idea because it makes files comparable between runs with Serial and Parallel Mesh. Defaults to True.
+
+- `partial`: Invokes a matching algorithm similar to the -m option.
```
```suggestion
- `partial`: invokes a matching algorithm similar to the -m option.
```

I'd suggest either always or never capitalizing

dschwen commented on modules/doc/content/python/testers/ImageDiff.md:
```diff
@@ -0,0 +1,30 @@
+
+`ImageDiff` tests compare images created by postprocessing of simulations. This is
+usually generated by a Python postprocessing script, but it would support images natively
+generated by an application.
```
```suggestion
generated by any application.
```
?

dschwen commented on modules/doc/content/python/testers/RunException.md:
```diff
@@ -0,0 +1,27 @@
+- should_crash: Indicates that the test is expected to crash or otherwise terminate early. Defaults to True
+
+
+Tests with an `expect_assert` parameter will only be ran in DEBUG and DEVEL mode, not in OPT mode.
```
```suggestion
Tests with an `expect_assert` parameter will only be run in DEBUG and DEVEL mode, not in OPT mode.
```

> 'will be' should be followed by a past participle. In this case the past participle of 'run' happens to be the same as the infinitive

dschwen commented on modules/doc/content/python/testers/XMLDiff.md:
```diff
@@ -0,0 +1,23 @@
+Test configuration options are added to the `tests` file.
+
+- `xmldiff`: A list of XML files to compare
+- `gold_dir`: The directory where the \"golden standard\" files reside relative to the TEST_DIR: (default: ./gold/)
```
```suggestion
- `gold_dir`: The directory where the \"gold standard\" files reside relative to the TEST_DIR: (default: ./gold/)
```

see https://pubmed.ncbi.nlm.nih.gov/16402524/

GiudGiud replied:
I copy pasted this from somewhere. i ll fix the source

dschwen commented on test/tests/reporters/perf_graph_reporter/tests:
```diff
@@ -14,7 +14,7 @@
       command = 'python verify_perf_graph_reporter.py'
       prereq = test/run
-      detail = 'and shall provide a tool to postproceses said information'
+      detail = 'and shall provide a tool to postprocess said information'
```
```suggestion
      detail = 'and shall provide a tool to post process said information'
```

dschwen commented on test/tests/reporters/perf_graph_reporter/tests:
```diff
@@ -41,7 +41,7 @@
       command = 'python verify_perf_graph_reporter.py recover'
       prereq = recover/run
-      detail = 'and shall provide a tool to postproceses said information'
+      detail = 'and shall provide a tool to postprocess said information'
```
```suggestion
      detail = 'and shall provide a tool to post process said information'
```

dschwen reviewed (dismissed):
Thanks for adding all those docs!

GiudGiud commented on modules/doc/content/python/testers/Exodiff.md:
```diff
@@ -0,0 +1,41 @@
+- `custom_cmp`: use a custom comparison. The output of the comparison
+  must match the specified file.
+
+- `map`: Use geometrical mapping to match up elements. This is usually a good idea because it makes files comparable between runs with Serial and Parallel Mesh. Defaults to True.
```
```suggestion
- `map`: use geometrical mapping to match up elements. This is usually a good idea because it makes files comparable between runs with Serial and Parallel Mesh. Defaults to True.
```

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Daniel Schwen <daniel@schwen.de>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/953686/) on aa3a113 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19985/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19985/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a0de0c877d2647785966c81bdfaaf03209e4e886`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/953831/) on 70e56a1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19985/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19985/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a0de0c877d2647785966c81bdfaaf03209e4e886`

GiudGiud commented:
@dschwen your comments should all be addressed

GiudGiud commented:
@dschwen the thumbs up emoji isnt enough for a merge :)

cticenhour commented on modules/doc/content/python/testers/AnalyzeJacobian.md:
```diff
@@ -0,0 +1,28 @@
+## Example test configuration in the MOOSE test suite
+
+In this example, three `AnalyzeJacobian` tests are created to check the Jacobian created by a finite difference
+preconditioner. The sucess of the Jacobian analysis is checked by looking for the string "No errors detected"
```
```suggestion
preconditioner. The success of the Jacobian analysis is checked by looking for the string "No errors detected"
```

cticenhour commented on modules/doc/content/python/testers/CSVDiff_tester.md:
```diff
@@ -0,0 +1,36 @@
+# CSVDiff
+
+`CSVDiff` tests compare `CSV` output(s) for the test to a reference in the in the specified
```
```suggestion
`CSVDiff` tests compare `CSV` output(s) for the test to a reference in the specified
```

cticenhour commented on modules/doc/content/python/testers/Exodiff.md:
```diff
@@ -0,0 +1,41 @@
+# Exodiff
```
Might be a good idea to mention in this page how to generate a custom cmp file:
```
${MOOSE_DIR}/framework/contrib/exodiff/exodiff -summary moose_simulation_out.e
```
and a listing of what one of those looks like. I don't remember seeing that anywhere in our current doco, and it would fit great here.

cticenhour replied:
Addressed. Thanks!

cticenhour commented on modules/doc/content/python/testers/Exodiff.md:
```diff
@@ -0,0 +1,41 @@
+
+- `map`: use geometrical mapping to match up elements. This is usually a good idea because it makes files comparable between runs with Serial and Parallel Mesh. Defaults to True.
+
+- `partial`: invokes a matching algorithm similar to the -m option.
```
```suggestion
- `partial`: invokes a matching algorithm similar to the `-m` option.
```

Also could you give some quick context for this option? Where is it used? I don't see it in an app executable or a run_tests one.

cticenhour commented on modules/doc/content/python/testers/RunException.md:
```diff
@@ -0,0 +1,27 @@
+
+Test configuration options are specified in the `tests` file.
+
+- expect_err: A regular expression or literal string that must occur in the output (see match_literal). (Test may terminate unexpectedly and be considered passing)
```
```suggestion
- `expect_err`: A regular expression or literal string that must occur in the output (see `match_literal`). (Test may terminate unexpectedly and be considered passing)
```

cticenhour commented on modules/doc/content/python/testers/RunException.md:
```diff
@@ -0,0 +1,27 @@
+
+- expect_err: A regular expression or literal string that must occur in the output (see match_literal). (Test may terminate unexpectedly and be considered passing)
+
+- expect_assert: DEBUG and DEVEL MODE ONLY: A regular expression that must occur in the output. (Test may terminate unexpectedly and be considered passing)
```
```suggestion
- `expect_assert`: DEBUG and DEVEL MODE ONLY: A regular expression that must occur in the output. (Test may terminate unexpectedly and be considered passing)
```

cticenhour commented on modules/doc/content/python/testers/RunException.md:
```diff
@@ -0,0 +1,27 @@
+
+- expect_assert: DEBUG and DEVEL MODE ONLY: A regular expression that must occur in the output. (Test may terminate unexpectedly and be considered passing)
+
+- should_crash: Indicates that the test is expected to crash or otherwise terminate early. Defaults to True
```
```suggestion
- `should_crash`: Indicates that the test is expected to crash or otherwise terminate early. Defaults to True
```

cticenhour commented on modules/doc/content/python/testers/XMLDiff.md:
```diff
@@ -0,0 +1,23 @@
+- `gold_dir`: The directory where the \"gold standard\" files reside relative to the TEST_DIR: (default: ./gold/)
+- `abs_zero`: Absolute zero cutoff used in exodiff comparisons, defaults to 1e-10
+- `rel_err`: Relative error value used in exodiff comparisons, defaults to 5.5e-6
+- `ignored_attributes`: Ignore e.g. type and/or version in sample XML block <VTKFile type=\"Foo\" version=\"0.1\">")
```
I would put that sample XML in the code tick marks for better viewing on the page

cticenhour commented on modules/doc/content/python/testers/JSONDiff.md:
```diff
@@ -0,0 +1,24 @@
+# JSONDiff
+
+`JSONDiff` tests compare `JSON` output(s) for the test to a reference in the in the specified
```
```suggestion
`JSONDiff` tests compare `JSON` output(s) for the test to a reference in the specified
```

cticenhour commented on modules/doc/content/python/testers/JSONDiff.md:
```diff
@@ -0,0 +1,24 @@
+# JSONDiff
+
+`JSONDiff` tests compare `JSON` output(s) for the test to a reference in the in the specified
+`gold_dir` folder. `JSON` output is generated by certain user objects, by the mesh  and reporters (for example,
```
```suggestion
`gold_dir` folder. `JSON` output is generated by certain user objects, by the mesh, and by reporters (for example,
```

cticenhour commented on modules/doc/content/python/testers/JSONDiff.md:
```diff
@@ -0,0 +1,24 @@
+
+Test configuration options are added to the `tests` file.
+
+- `jsondiff`: A list of JSON files to compare
```
```suggestion
- `jsondiff`: A list of `JSON` files to compare
```

cticenhour commented on modules/doc/content/python/testers/JSONDiff.md:
```diff
@@ -0,0 +1,24 @@
+
+- `jsondiff`: A list of JSON files to compare
+
+- `skip_keys`: A list of keys to skip in the JSON comparison
```
```suggestion
- `skip_keys`: A list of keys to skip in the `JSON` comparison
```

cticenhour commented on modules/doc/content/python/testers/JSONDiff.md:
```diff
@@ -0,0 +1,24 @@
+## Example test configuration in the MOOSE test suite
+
+In this example, the mesh output by a [RenameBlockGenerator.md] is being checked using a
+`MeshInfo` reporter, which outputs a JSON file. This JSON file is compared to a reference to
```
```suggestion
`MeshInfo` reporter, which outputs a `JSON` file. This `JSON` file is compared to a reference to
```

cticenhour commented on modules/doc/content/python/testers/PetscJacobianTester.md:
```diff
@@ -0,0 +1,35 @@
+# PetscJacobianTester
+
+The `PetscJacobianTester` tester uses the `-snes_test_jacobian` command line option
+in Petsc to evaluate the quality of the Jacobian in the specified tests.
```
```suggestion
in PETSc to evaluate the quality of the Jacobian in the specified tests.
```

cticenhour commented on modules/doc/content/python/testers/PythonUnitTest.md:
```diff
@@ -0,0 +1,34 @@
+
+- `test_case`: The specific test case to run. Defaults to all test cases in the module
+
+- `buffer`: Equivalent to passing -b or --buffer to the unittest. Defaults to False
```
```suggestion
- `buffer`: Equivalent to passing `-b` or `--buffer` to the unittest. Defaults to False
```

cticenhour commented on modules/doc/content/python/TestHarness.md:
```diff
@@ -75,18 +75,22 @@ Testers represent individual tests in MOOSE. Testers encompass a set of instruct
+- [JSONDiff](JSONDiff.md)
+- [XMLDiff](XMLDiff.md)
+- [ImageDiff](ImageDiff.md)
 - [CSVValidationTester](CSVValidationTester.md)
-- FileTester
-- AnalyzeJacobian
-- CheckFiles
-- ImageDiff
-- PetscJacobianTester
-- RunCommand
-- RunException
-- VTKDiff
+- [FileTester](FileTester.md)
+- [CheckFiles](CheckFiles.md)
+- [AnalyzeJacobian](AnalyzeJacobian.md)
+- [PetscJacobianTester](PetscJacobianTester.md)
+- [PythonUnitTest](PythonUnitTest.md) (includes [Method of Manufactured solutions](python/mms.md) testing)
+- [BenchmarkTesting](application_development/performance_benchmarking.md)
```
Thanks for adding all of this!

GiudGiud committed:
Apply suggestions from code review
Remove unacceptable petsc capitalization

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

GiudGiud committed:
Add examples and text for creating custom exodus comparisons

Update modules/doc/content/python/testers/Exodiff.md

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

Update modules/doc/content/python/testers/Exodiff.md

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

cticenhour commented on modules/doc/content/python/testers/Exodiff.md:
```diff
@@ -0,0 +1,66 @@
+  may be found in its [documentation](https://gsjaardema.github.io/seacas-docs/sphinx/html/index.html#exodiff)
+
+- `custom_cmp`: use a custom comparison. The output of the comparison
+  must match the specified file. See [below](custom_cmp) for more details
```
```suggestion
  must match the specified file. See [below](#custom_cmp) for more details
```
This link was broken.

cticenhour commented on modules/doc/content/python/testers/Exodiff.md:
````diff
@@ -0,0 +1,66 @@
+${MOOSE_DIR}/framework/contrib/exodiff/exodiff -summary moose_simulation_out.e
+```
+
+This generates a summary of the simulation, which can be edited to make the
````
```suggestion
This generates a summary of the simulation which can be edited to make the
```

cticenhour reviewed (changes requested):
Only two more things, since I found a broken link, and this is ready to go.
